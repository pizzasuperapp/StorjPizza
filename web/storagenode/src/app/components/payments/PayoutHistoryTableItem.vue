// Copyright (C) 2020 Storj Labs, Inc.
// See LICENSE for copying information.

<template>
    <div class="payout-history-item">
        <div class="payout-history-item__header" @click="toggleExpanded">
            <div class="payout-history-item__header__left-area">
                <div class="payout-history-item__header__left-area__expand-icon">
                    <ExpandIcon />
                </div>
                <p>{{ historyItem.satelliteName }}</p>
            </div>
            <p class="payout-history-item__header__total">{{ historyItem.paid | centsToDollars }}</p>
        </div>
        <transition name="fade" mode="in-out">
            <div v-if="isExpanded" class="payout-history-item__expanded-area">
                <div class="payout-history-item__expanded-area__left-area">
                    <div class="payout-history-item__expanded-area__left-area__info-area">
                        <div class="payout-history-item__expanded-area__left-area__info-area__item flex-start">
                            <p class="payout-history-item__expanded-area__left-area__info-area__item__label extra-margin">Node Age</p>
                            <p class="payout-history-item__expanded-area__left-area__info-area__item__value">{{ `${historyItem.age} Month${historyItem.age > 1 ? 's' : ''}` }}</p>
                        </div>
                        <div class="payout-history-item__expanded-area__left-area__info-area__item flex-start">
                            <p class="payout-history-item__expanded-area__left-area__info-area__item__label extra-margin">Earned</p>
                            <p class="payout-history-item__expanded-area__left-area__info-area__item__value">{{ historyItem.earned | centsToDollars }}</p>
                        </div>
                        <div class="payout-history-item__expanded-area__left-area__info-area__item flex-end">
                            <div class="row extra-margin">
                                <p class="payout-history-item__expanded-area__left-area__info-area__item__label">Surge</p>
                                <div class="payout-history-item__expanded-area__left-area__info-area__item__info-block">
                                    <p class="payout-history-item__expanded-area__left-area__info-area__item__info-block__text">{{ historyItem.surgePercent + '%' }}</p>
                                </div>
                            </div>
                            <p class="payout-history-item__expanded-area__left-area__info-area__item__value">{{ historyItem.surge | centsToDollars }}</p>
                        </div>
                        <div class="payout-history-item__expanded-area__left-area__info-area__item flex-end">
                            <div class="row extra-margin">
                                <p class="payout-history-item__expanded-area__left-area__info-area__item__label">Held</p>
                                <div class="payout-history-item__expanded-area__left-area__info-area__item__info-block">
                                    <p class="payout-history-item__expanded-area__left-area__info-area__item__info-block__text">{{ historyItem.heldPercent + '%' }}</p>
                                </div>
                            </div>
                            <p class="payout-history-item__expanded-area__left-area__info-area__item__value">{{ historyItem.held | centsToDollars }}</p>
                        </div>
                    </div>
                    <div class="payout-history-item__expanded-area__left-area__footer">
                        <div v-if="historyItem.isExitComplete" class="payout-history-item__expanded-area__left-area__footer__item">
                            <DisqualifyIcon class="disqualify-icon" />
                            <div class="payout-history-item__expanded-area__left-area__footer__item__text-area">
                                <p class="payout-history-item__expanded-area__left-area__footer__item__text-area__title">
                                    Your node made Graceful Exit
                                </p>
                                <p class="payout-history-item__expanded-area__left-area__footer__item__text-area__text">
                                    100% of total withholdings are returned
                                </p>
                            </div>
                        </div>
                        <div v-if="historyItem.age > 9 && !historyItem.isExitComplete" class="payout-history-item__expanded-area__left-area__footer__item">
                            <OKIcon class="ok-icon" />
                            <div class="payout-history-item__expanded-area__left-area__footer__item__text-area">
                                <p class="payout-history-item__expanded-area__left-area__footer__item__text-area__title">
                                    Your node reached age of 10 Month
                                </p>
                                <p class="payout-history-item__expanded-area__left-area__footer__item__text-area__text">
                                    100% of storage node revenue is paid
                                </p>
                            </div>
                        </div>
                    </div>
                </div>
                <div class="payout-history-item__expanded-area__right-area">
                    <p class="payout-history-item__expanded-area__right-area__label flex-end">Paid</p>
                    <div class="payout-history-item__expanded-area__right-area__info-item">
                        <p class="payout-history-item__expanded-area__right-area__info-item__label">After Held</p>
                        <p class="payout-history-item__expanded-area__right-area__info-item__value">{{ historyItem.afterHeld | centsToDollars }}</p>
                    </div>
                    <div class="payout-history-item__expanded-area__right-area__info-item">
                        <p class="payout-history-item__expanded-area__right-area__info-item__label">Held Returned</p>
                        <p class="payout-history-item__expanded-area__right-area__info-item__value">{{ historyItem.disposed | centsToDollars }}</p>
                    </div>
                    <div class="payout-history-item__expanded-area__right-area__info-item">
                        <p class="payout-history-item__expanded-area__right-area__info-item__label">Distributed</p>
                        <p class="payout-history-item__expanded-area__right-area__info-item__value">{{ historyItem.distributed | centsToDollars }}</p>
                    </div>
                    <div class="payout-history-item__expanded-area__right-area__divider" />
                    <div class="payout-history-item__expanded-area__right-area__footer">
                        <div v-if="historyItem.transactionLink" class="payout-history-item__expanded-area__right-area__footer__transaction">
                            <a
                                class="payout-history-item__expanded-area__right-area__footer__transaction__link"
                                :href="historyItem.transactionLink"
                                target="_blank"
                                rel="noreferrer noopener"
                            >
                                Transaction
                            </a>
                            <ShareIcon class="payout-history-item__expanded-area__right-area__footer__transaction__icon" />
                        </div>
                        <p class="payout-history-item__expanded-area__right-area__footer__total">{{ (historyItem.distributed > 0 ? historyItem.distributed : historyItem.paid) | centsToDollars }}</p>
                    </div>
                </div>
            </div>
        </transition>
    </div>
</template>

<script lang="ts">
import { Component, Prop, Vue } from 'vue-property-decorator';

import ExpandIcon from '@/../static/images/BlueArrowRight.svg';
import DisqualifyIcon from '@/../static/images/largeDisqualify.svg';
import OKIcon from '@/../static/images/payments/OKIcon.svg';
import ShareIcon from '@/../static/images/payments/Share.svg';

import { SatellitePayoutForPeriod } from '@/storagenode/payouts/payouts';

// @vue/component
@Component ({
    components: {
        ExpandIcon,
        ShareIcon,
        DisqualifyIcon,
        OKIcon,
    },
})
export default class PayoutHistoryTableItem extends Vue {
    @Prop({default: () => new SatellitePayoutForPeriod()})
    public readonly historyItem: SatellitePayoutForPeriod;

    /**
     * Indicates if payout info should be rendered.
     */
    public isExpanded = false;

    /**
     * Toggles additional payout information.
     */
    public toggleExpanded(): void {
        this.isExpanded = !this.isExpanded;
    }
}
</script>

<style scoped lang="scss">
    .payout-history-item {
        padding: 17px;
        width: calc(100% - 34px);
        border-bottom: 1px solid #eaeaea;

        &__header {
            display: flex;
            justify-content: space-between;
            align-items: center;
            font-family: 'font_medium', sans-serif;
            font-size: 14px;
            color: var(--regular-text-color);
            cursor: pointer;

            &__left-area {
                display: flex;
                align-items: center;
                justify-content: flex-start;

                &__expand-icon {
                    display: flex;
                    align-items: center;
                    justify-content: center;
                    width: 14px;
                    height: 14px;
                    max-width: 14px;
                    max-height: 14px;
                    margin-right: 12px;
                }
            }
        }

        &__expanded-area {
            position: relative;
            display: flex;
            align-items: flex-start;
            justify-content: space-between;
            padding-left: 26px;
            width: calc(100% - 26px);
            margin-top: 20px;

            &__left-area {
                display: flex;
                flex-direction: column;
                justify-content: flex-start;
                align-items: flex-start;
                width: 65%;

                &__info-area {
                    display: flex;
                    align-items: center;
                    justify-content: space-between;
                    width: 100%;

                    &__item {
                        display: flex;
                        flex-direction: column;
                        font-family: 'font_medium', sans-serif;
                        font-size: 14px;

                        &__label {
                            color: var(--label-text-color);
                        }

                        &__value {
                            color: var(--regular-text-color);
                        }

                        &__info-block {
                            margin-left: 5px;
                            border-radius: 4px;
                            background: #909bad;
                            padding: 3px;

                            &__text {
                                font-size: 11px;
                                line-height: 11px;
                                color: white;
                            }
                        }
                    }
                }

                &__footer {
                    margin-top: 30px;
                    width: 100%;

                    &__item {
                        display: flex;
                        justify-content: flex-start;
                        align-items: center;

                        &__text-area {
                            display: flex;
                            flex-direction: column;
                            align-items: flex-start;
                            justify-content: center;
                            margin-left: 12px;

                            &__title {
                                font-family: 'font_bold', sans-serif;
                                font-size: 15px;
                                color: var(--regular-text-color);
                            }

                            &__text {
                                margin-top: 6px;
                                font-family: 'font_regular', sans-serif;
                                font-size: 13px;
                                color: var(--regular-text-color);
                            }
                        }
                    }
                }
            }

            &__right-area {
                width: 35%;
                display: flex;
                flex-direction: column;
                font-family: 'font_medium', sans-serif;
                font-size: 14px;

                &__label {
                    color: var(--label-text-color);
                    text-align: end;
                }

                &__info-item {
                    display: flex;
                    align-items: center;
                    justify-content: flex-end;

                    &__label,
                    &__value {
                        margin-top: 10px;
                        color: var(--regular-text-color);
                    }

                    &__value {
                        margin-left: 30px;
                        width: 60px;
                        text-align: end;
                    }
                }

                &__divider {
                    width: 100%;
                    height: 1px;
                    background: #eaeaea;
                    margin: 15px 0;
                }

                &__footer {
                    display: flex;
                    align-items: flex-end;
                    justify-content: flex-end;

                    &__transaction {
                        display: flex;
                        align-items: center;
                        justify-content: flex-end;
                        color: var(--navigation-link-color);
                        cursor: pointer;

                        a:visited {
                            color: var(--navigation-link-color);
                        }

                        &__icon {
                            margin-left: 7px;

                            ::v-deep path {
                                stroke: var(--navigation-link-color);
                            }
                        }
                    }

                    &__total {
                        margin-left: 30px;
                        width: 60px;
                        font-family: 'font_bold', sans-serif;
                        color: var(--regular-text-color);
                        text-align: end;
                    }
                }
            }
        }
    }

    .row {
        display: flex;
        align-items: flex-end;
        justify-content: center;
    }

    .extra-margin {
        margin-bottom: 8px;
    }

    .flex-start {
        justify-content: flex-start;
        align-items: flex-start;
    }

    .flex-end {
        justify-content: flex-end;
        align-items: flex-end;
    }

    .fade-enter-active,
    .fade-leave-active {
        transition: opacity 0.5s;
    }

    .fade-enter,
    .fade-leave-to {
        opacity: 0;
    }

    .disqualify-icon {
        width: 40px;
        height: 40px;
        min-width: 40px;
        min-height: 40px;

        ::v-deep path {
            fill: #909bad;
        }
    }

    .ok-icon {
        width: 30px;
        height: 30px;
        min-width: 30px;
        min-height: 30px;
    }

    @media screen and (max-width: 800px) {

        .payout-history-item {
            padding: 17px 10px 12px;
            width: calc(100% - 20px);

            &__header {
                flex-wrap: wrap;

                &__left-area {
                    flex-wrap: nowrap;
                    margin-bottom: 5px;
                }

                &__total {
                    margin-bottom: 5px;
                }
            }

            &__expanded-area {
                width: 100%;
                padding: 0;
                flex-direction: column;

                &__left-area {
                    width: 100%;

                    &__info-area {
                        flex-wrap: wrap;
                        width: 100%;

                        &__item {
                            align-items: flex-start;
                            justify-content: flex-start;
                            margin-bottom: 5px;
                        }
                    }

                    &__footer {
                        width: 100%;
                    }
                }

                &__right-area {
                    width: 100%;
                }
            }
        }
    }
</style>
