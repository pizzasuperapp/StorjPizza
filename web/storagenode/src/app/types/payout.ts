// Copyright (C) 2020 Storj Labs, Inc.
// See LICENSE for copying information.

import {
    EstimatedPayout,
    PayoutPeriod,
    SatelliteHeldHistory,
    SatellitePayoutForPeriod,
    TotalPayments,
    TotalPaystubForPeriod,
} from '@/storagenode/payouts/payouts';

/**
 * Holds 'start' and 'end' of payout period range.
 */
export class PayoutInfoRange {
    public constructor(
        public start: PayoutPeriod | null = null,
        public end: PayoutPeriod = new PayoutPeriod(),
    ) {}
}

/**
 * Holds all payout module state.
 */
export class PayoutState {
    public constructor (
        public totalPaystubForPeriod: TotalPaystubForPeriod = new TotalPaystubForPeriod(),
        public periodRange: PayoutInfoRange = new PayoutInfoRange(),
        public totalPayments: TotalPayments = new TotalPayments(),
        public currentMonthEarnings: number = 0,
        public heldPercentage: number = 0,
        public payoutPeriods: PayoutPeriod[] = [],
        public heldHistory: SatelliteHeldHistory[] = [],
        public payoutHistory: SatellitePayoutForPeriod[] = [],
        public payoutHistoryPeriod: string = '',
        public estimation: EstimatedPayout = new EstimatedPayout(),
        public payoutHistoryAvailablePeriods: PayoutPeriod[] = [],
    ) {}
}

export interface StoredMonthsByYear {
    [key: number]: MonthButton[];
}

/**
 * Holds all months names.
 */
export const monthNames = [
    'January', 'February', 'March', 'April',
    'May', 'June', 'July',	'August',
    'September', 'October', 'November',	'December',
];

/**
 * Describes month button entity for calendar.
 */
export class MonthButton {
    public constructor(
        public year: number = 0,
        public index: number = 0,
        public active: boolean = false,
        public selected: boolean = false,
    ) {}

    /**
     * Returns month label depends on index.
     */
    public get name(): string {
        return monthNames[this.index] ? monthNames[this.index].slice(0, 3) : '';
    }
}
