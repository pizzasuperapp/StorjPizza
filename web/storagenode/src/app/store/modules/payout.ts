// Copyright (C) 2020 Storj Labs, Inc.
// See LICENSE for copying information.

import { StoreModule } from '@/app/store';
import {
    PayoutInfoRange,
    PayoutState,
} from '@/app/types/payout';
import { StorageNodeState } from '@/app/types/sno';
import { getHeldPercentage } from '@/app/utils/payout';
import { SizeBreakpoints } from '@/private/memory/size';
import {
    EstimatedPayout,
    PayoutPeriod,
    SatelliteHeldHistory,
    SatellitePayoutForPeriod,
    TotalPayments,
    TotalPaystubForPeriod,
} from '@/storagenode/payouts/payouts';
import { PayoutService } from '@/storagenode/payouts/service';

export const PAYOUT_MUTATIONS = {
    SET_PAYOUT_INFO: 'SET_PAYOUT_INFO',
    SET_RANGE: 'SET_RANGE',
    SET_TOTAL: 'SET_TOTAL',
    SET_HELD_PERCENT: 'SET_HELD_PERCENT',
    SET_HELD_HISTORY: 'SET_HELD_HISTORY',
    SET_ESTIMATION: 'SET_ESTIMATION',
    SET_PERIODS: 'SET_PERIODS',
    SET_PAYOUT_HISTORY: 'SET_PAYOUT_HISTORY',
    SET_PAYOUT_HISTORY_PERIOD: 'SET_PAYOUT_HISTORY_PERIOD',
    SET_PAYOUT_HISTORY_AVAILABLE_PERIODS: 'SET_PAYOUT_HISTORY_AVAILABLE_PERIODS',
};

export const PAYOUT_ACTIONS = {
    GET_PAYOUT_INFO: 'GET_PAYOUT_INFO',
    SET_PERIODS_RANGE: 'SET_PERIODS_RANGE',
    GET_TOTAL: 'GET_TOTAL',
    GET_HELD_HISTORY: 'GET_HELD_HISTORY',
    GET_ESTIMATION: 'GET_ESTIMATION',
    GET_PERIODS: 'GET_PERIODS',
    GET_PAYOUT_HISTORY: 'GET_PAYOUT_HISTORY',
    SET_PAYOUT_HISTORY_PERIOD: 'SET_PAYOUT_HISTORY_PERIOD',
};

// TODO: move to config in storagenode/payouts
export const BANDWIDTH_DOWNLOAD_PRICE_PER_TB = 2000;
export const BANDWIDTH_REPAIR_PRICE_PER_TB = 1000;
export const DISK_SPACE_PRICE_PER_TB = 150;

interface PayoutContext {
    rootState: {
        node: StorageNodeState;
    };
    state: PayoutState;
    commit: (string, ...unknown) => void;
}

/**
 * creates notifications module with all dependencies
 *
 * @param service - payments service
 */
export function newPayoutModule(service: PayoutService): StoreModule<PayoutState> {
    return {
        state: new PayoutState(),
        mutations: {
            [PAYOUT_MUTATIONS.SET_PAYOUT_INFO](state: PayoutState, totalPaystubForPeriod: TotalPaystubForPeriod): void {
                state.totalPaystubForPeriod = totalPaystubForPeriod;
            },
            [PAYOUT_MUTATIONS.SET_TOTAL](state: PayoutState, totalPayments: TotalPayments): void {
                state.totalPayments = totalPayments;
                state.currentMonthEarnings = totalPayments.currentMonthEarnings;
            },
            [PAYOUT_MUTATIONS.SET_RANGE](state: PayoutState, periodRange: PayoutInfoRange): void {
                state.periodRange = periodRange;
            },
            [PAYOUT_MUTATIONS.SET_HELD_PERCENT](state: PayoutState, heldPercentage: number): void {
                state.heldPercentage = heldPercentage;
            },
            [PAYOUT_MUTATIONS.SET_HELD_HISTORY](state: PayoutState, heldHistory: SatelliteHeldHistory[]): void {
                state.heldHistory = heldHistory;
            },
            [PAYOUT_MUTATIONS.SET_ESTIMATION](state: PayoutState, estimatedInfo: EstimatedPayout): void {
                state.estimation = estimatedInfo;
            },
            [PAYOUT_MUTATIONS.SET_PERIODS](state: PayoutState, periods: PayoutPeriod[]): void {
                state.payoutPeriods = periods;
            },
            [PAYOUT_MUTATIONS.SET_PAYOUT_HISTORY](state: PayoutState, payoutHistory: SatellitePayoutForPeriod[]): void {
                state.payoutHistory = payoutHistory;
            },
            [PAYOUT_MUTATIONS.SET_PAYOUT_HISTORY_PERIOD](state: PayoutState, period: string): void {
                state.payoutHistoryPeriod = period;
            },
            [PAYOUT_MUTATIONS.SET_PAYOUT_HISTORY_AVAILABLE_PERIODS](state: PayoutState, periods: PayoutPeriod[]): void {
                state.payoutHistoryAvailablePeriods = periods;
            },
        },
        actions: {
            [PAYOUT_ACTIONS.GET_PAYOUT_INFO]: async function ({ commit, state, rootState }: PayoutContext, satelliteId = ''): Promise<void> {
                const totalPaystubForPeriod = await service.paystubSummaryForPeriod(
                    state.periodRange.start,
                    state.periodRange.end,
                    satelliteId,
                );

                commit(PAYOUT_MUTATIONS.SET_HELD_PERCENT, getHeldPercentage(rootState.node.selectedSatellite.joinDate));
                commit(PAYOUT_MUTATIONS.SET_PAYOUT_INFO, totalPaystubForPeriod);
            },
            [PAYOUT_ACTIONS.GET_TOTAL]: async function ({ commit, rootState }: PayoutContext, satelliteId = ''): Promise<void> {
                const now = new Date();
                const start = new PayoutPeriod(rootState.node.selectedSatellite.joinDate.getUTCFullYear(), rootState.node.selectedSatellite.joinDate.getUTCMonth());
                const end = new PayoutPeriod(now.getUTCFullYear(), now.getUTCMonth());

                const totalPayments = await service.totalPayments(start, end, satelliteId);

                // TODO: move to service
                const currentBandwidthDownload = (rootState.node.egressChartData || [])
                    .map(data => data.egress.usage)
                    .reduce((previous, current) => previous + current, 0);

                const currentBandwidthAuditAndRepair = (rootState.node.egressChartData || [])
                    .map(data => data.egress.audit + data.egress.repair)
                    .reduce((previous, current) => previous + current, 0);

                const approxHourInMonth = 720;
                const currentDiskSpace = (rootState.node.storageChartData || [])
                    .map(data => data.atRestTotal)
                    .reduce((previous, current) => previous + current, 0) / approxHourInMonth;

                const thisMonthEarnings = (currentBandwidthDownload * BANDWIDTH_DOWNLOAD_PRICE_PER_TB
                    + currentBandwidthAuditAndRepair * BANDWIDTH_REPAIR_PRICE_PER_TB
                    + currentDiskSpace * DISK_SPACE_PRICE_PER_TB) / SizeBreakpoints.TB;

                totalPayments.setCurrentMonthEarnings(thisMonthEarnings);

                commit(PAYOUT_MUTATIONS.SET_HELD_PERCENT, getHeldPercentage(rootState.node.selectedSatellite.joinDate));
                commit(PAYOUT_MUTATIONS.SET_TOTAL, totalPayments);
            },
            [PAYOUT_ACTIONS.SET_PERIODS_RANGE]: function ({ commit }: PayoutContext, periodRange: PayoutInfoRange): void {
                commit(PAYOUT_MUTATIONS.SET_RANGE, periodRange);
            },
            [PAYOUT_ACTIONS.GET_HELD_HISTORY]: async function ({ commit }: PayoutContext): Promise<void> {
                const heldHistory = await service.allSatellitesHeldHistory();

                commit(PAYOUT_MUTATIONS.SET_HELD_HISTORY, heldHistory);
            },
            [PAYOUT_ACTIONS.GET_PERIODS]: async function ({commit}: PayoutContext, satelliteId = ''): Promise<void> {
                const periods = await service.availablePeriods(satelliteId);

                commit(PAYOUT_MUTATIONS.SET_PERIODS, periods);

                if (!satelliteId) {
                    commit(PAYOUT_MUTATIONS.SET_PAYOUT_HISTORY_AVAILABLE_PERIODS, periods);
                }
            },
            [PAYOUT_ACTIONS.GET_ESTIMATION]: async function ({ commit }: PayoutContext, satelliteId = ''): Promise<void> {
                const estimatedInfo = await service.estimatedPayout(satelliteId);

                commit(PAYOUT_MUTATIONS.SET_ESTIMATION, estimatedInfo);
            },
            [PAYOUT_ACTIONS.GET_PAYOUT_HISTORY]: async function ({ commit, state }: PayoutContext): Promise<void> {
                if (!state.payoutHistoryPeriod) return;

                const payoutHistory = await service.payoutHistory(state.payoutHistoryPeriod);

                commit(PAYOUT_MUTATIONS.SET_PAYOUT_HISTORY, payoutHistory);
            },
            [PAYOUT_ACTIONS.SET_PAYOUT_HISTORY_PERIOD]: function ({ commit }: PayoutContext, period: string): void {
                commit(PAYOUT_MUTATIONS.SET_PAYOUT_HISTORY_PERIOD, period);
            },
        },
        getters: {
            totalPaidForPayoutHistoryPeriod: (state: PayoutState): number => {
                return state.payoutHistory.map(data => data.paid)
                    .reduce((previous, current) => previous + current, 0);
            },
        },
    };
}
