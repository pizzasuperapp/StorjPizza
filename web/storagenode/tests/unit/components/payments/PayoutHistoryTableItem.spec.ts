// Copyright (C) 2020 Storj Labs, Inc.
// See LICENSE for copying information.

import Vuex from 'vuex';

import PayoutHistoryTableItem from '@/app/components/payments/PayoutHistoryTableItem.vue';

import { newPayoutModule } from '@/app/store/modules/payout';
import { PayoutHttpApi } from '@/storagenode/api/payout';
import { SatellitePayoutForPeriod } from '@/storagenode/payouts/payouts';
import { PayoutService } from '@/storagenode/payouts/service';
import { createLocalVue, shallowMount } from '@vue/test-utils';

const localVue = createLocalVue();
localVue.use(Vuex);

localVue.filter('centsToDollars', (cents: number): string => {
    return `$${(cents / 100).toFixed(2)}`;
});

const payoutApi = new PayoutHttpApi();
const payoutService = new PayoutService(payoutApi);
const payoutModule = newPayoutModule(payoutService);

const store = new Vuex.Store({ modules: { payoutModule }});

describe('PayoutHistoryTableItem', (): void => {
    it('renders correctly with actual values (eth)', async (): Promise<void> => {
        const wrapper = shallowMount(PayoutHistoryTableItem, {
            store,
            localVue,
            propsData: {
                historyItem: new SatellitePayoutForPeriod('1', 'name1', 2, 100000, 1200000, 140,
                    500000, 600000, 200000, 800000, 'eth:receipt1', false, 75, 300000,
                ),
            },
        });

        expect(wrapper).toMatchSnapshot();

        await wrapper.find('.payout-history-item__header').trigger('click');

        expect(wrapper.vm.$data.isExpanded).toBe(true);
        expect(wrapper).toMatchSnapshot();
    });

    it('renders correctly with actual values (zksync)', async (): Promise<void> => {
        const wrapper = shallowMount(PayoutHistoryTableItem, {
            store,
            localVue,
            propsData: {
                historyItem: new SatellitePayoutForPeriod('1', 'name1', 1, 100000, 1200000, 140,
                    500000, 600000, 200000, 800000, 'zksync:receipt1', false, 75, 300000,
                ),
            },
        });

        await wrapper.find('.payout-history-item__header').trigger('click');
        expect(wrapper).toMatchSnapshot();
    });
});
