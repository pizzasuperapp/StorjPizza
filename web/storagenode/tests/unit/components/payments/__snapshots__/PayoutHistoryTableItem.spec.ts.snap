// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PayoutHistoryTableItem renders correctly with actual values (eth) 1`] = `
<div class="payout-history-item">
  <div class="payout-history-item__header">
    <div class="payout-history-item__header__left-area">
      <div class="payout-history-item__header__left-area__expand-icon">
        <expandicon-stub></expandicon-stub>
      </div>
      <p>name1</p>
    </div>
    <p class="payout-history-item__header__total">$0.80</p>
  </div>
  <transition-stub name="fade" mode="in-out">
    <!---->
  </transition-stub>
</div>
`;

exports[`PayoutHistoryTableItem renders correctly with actual values (eth) 2`] = `
<div class="payout-history-item">
  <div class="payout-history-item__header">
    <div class="payout-history-item__header__left-area">
      <div class="payout-history-item__header__left-area__expand-icon">
        <expandicon-stub></expandicon-stub>
      </div>
      <p>name1</p>
    </div>
    <p class="payout-history-item__header__total">$0.80</p>
  </div>
  <transition-stub name="fade" mode="in-out">
    <div class="payout-history-item__expanded-area">
      <div class="payout-history-item__expanded-area__left-area">
        <div class="payout-history-item__expanded-area__left-area__info-area">
          <div class="payout-history-item__expanded-area__left-area__info-area__item flex-start">
            <p class="payout-history-item__expanded-area__left-area__info-area__item__label extra-margin">Node Age</p>
            <p class="payout-history-item__expanded-area__left-area__info-area__item__value">2 Months</p>
          </div>
          <div class="payout-history-item__expanded-area__left-area__info-area__item flex-start">
            <p class="payout-history-item__expanded-area__left-area__info-area__item__label extra-margin">Earned</p>
            <p class="payout-history-item__expanded-area__left-area__info-area__item__value">$0.10</p>
          </div>
          <div class="payout-history-item__expanded-area__left-area__info-area__item flex-end">
            <div class="row extra-margin">
              <p class="payout-history-item__expanded-area__left-area__info-area__item__label">Surge</p>
              <div class="payout-history-item__expanded-area__left-area__info-area__item__info-block">
                <p class="payout-history-item__expanded-area__left-area__info-area__item__info-block__text">140%</p>
              </div>
            </div>
            <p class="payout-history-item__expanded-area__left-area__info-area__item__value">$1.20</p>
          </div>
          <div class="payout-history-item__expanded-area__left-area__info-area__item flex-end">
            <div class="row extra-margin">
              <p class="payout-history-item__expanded-area__left-area__info-area__item__label">Held</p>
              <div class="payout-history-item__expanded-area__left-area__info-area__item__info-block">
                <p class="payout-history-item__expanded-area__left-area__info-area__item__info-block__text">75%</p>
              </div>
            </div>
            <p class="payout-history-item__expanded-area__left-area__info-area__item__value">$0.50</p>
          </div>
        </div>
        <div class="payout-history-item__expanded-area__left-area__footer">
          <!---->
          <!---->
        </div>
      </div>
      <div class="payout-history-item__expanded-area__right-area">
        <p class="payout-history-item__expanded-area__right-area__label flex-end">Paid</p>
        <div class="payout-history-item__expanded-area__right-area__info-item">
          <p class="payout-history-item__expanded-area__right-area__info-item__label">After Held</p>
          <p class="payout-history-item__expanded-area__right-area__info-item__value">$0.60</p>
        </div>
        <div class="payout-history-item__expanded-area__right-area__info-item">
          <p class="payout-history-item__expanded-area__right-area__info-item__label">Held Returned</p>
          <p class="payout-history-item__expanded-area__right-area__info-item__value">$0.20</p>
        </div>
        <div class="payout-history-item__expanded-area__right-area__info-item">
          <p class="payout-history-item__expanded-area__right-area__info-item__label">Distributed</p>
          <p class="payout-history-item__expanded-area__right-area__info-item__value">$0.30</p>
        </div>
        <div class="payout-history-item__expanded-area__right-area__divider"></div>
        <div class="payout-history-item__expanded-area__right-area__footer">
          <div class="payout-history-item__expanded-area__right-area__footer__transaction"><a href="https://etherscan.io/tx/0xreceipt1" target="_blank" rel="noreferrer noopener" class="payout-history-item__expanded-area__right-area__footer__transaction__link">
              Transaction
            </a>
            <shareicon-stub class="payout-history-item__expanded-area__right-area__footer__transaction__icon"></shareicon-stub>
          </div>
          <p class="payout-history-item__expanded-area__right-area__footer__total">$0.30</p>
        </div>
      </div>
    </div>
  </transition-stub>
</div>
`;

exports[`PayoutHistoryTableItem renders correctly with actual values (zksync) 1`] = `
<div class="payout-history-item">
  <div class="payout-history-item__header">
    <div class="payout-history-item__header__left-area">
      <div class="payout-history-item__header__left-area__expand-icon">
        <expandicon-stub></expandicon-stub>
      </div>
      <p>name1</p>
    </div>
    <p class="payout-history-item__header__total">$0.80</p>
  </div>
  <transition-stub name="fade" mode="in-out">
    <div class="payout-history-item__expanded-area">
      <div class="payout-history-item__expanded-area__left-area">
        <div class="payout-history-item__expanded-area__left-area__info-area">
          <div class="payout-history-item__expanded-area__left-area__info-area__item flex-start">
            <p class="payout-history-item__expanded-area__left-area__info-area__item__label extra-margin">Node Age</p>
            <p class="payout-history-item__expanded-area__left-area__info-area__item__value">1 Month</p>
          </div>
          <div class="payout-history-item__expanded-area__left-area__info-area__item flex-start">
            <p class="payout-history-item__expanded-area__left-area__info-area__item__label extra-margin">Earned</p>
            <p class="payout-history-item__expanded-area__left-area__info-area__item__value">$0.10</p>
          </div>
          <div class="payout-history-item__expanded-area__left-area__info-area__item flex-end">
            <div class="row extra-margin">
              <p class="payout-history-item__expanded-area__left-area__info-area__item__label">Surge</p>
              <div class="payout-history-item__expanded-area__left-area__info-area__item__info-block">
                <p class="payout-history-item__expanded-area__left-area__info-area__item__info-block__text">140%</p>
              </div>
            </div>
            <p class="payout-history-item__expanded-area__left-area__info-area__item__value">$1.20</p>
          </div>
          <div class="payout-history-item__expanded-area__left-area__info-area__item flex-end">
            <div class="row extra-margin">
              <p class="payout-history-item__expanded-area__left-area__info-area__item__label">Held</p>
              <div class="payout-history-item__expanded-area__left-area__info-area__item__info-block">
                <p class="payout-history-item__expanded-area__left-area__info-area__item__info-block__text">75%</p>
              </div>
            </div>
            <p class="payout-history-item__expanded-area__left-area__info-area__item__value">$0.50</p>
          </div>
        </div>
        <div class="payout-history-item__expanded-area__left-area__footer">
          <!---->
          <!---->
        </div>
      </div>
      <div class="payout-history-item__expanded-area__right-area">
        <p class="payout-history-item__expanded-area__right-area__label flex-end">Paid</p>
        <div class="payout-history-item__expanded-area__right-area__info-item">
          <p class="payout-history-item__expanded-area__right-area__info-item__label">After Held</p>
          <p class="payout-history-item__expanded-area__right-area__info-item__value">$0.60</p>
        </div>
        <div class="payout-history-item__expanded-area__right-area__info-item">
          <p class="payout-history-item__expanded-area__right-area__info-item__label">Held Returned</p>
          <p class="payout-history-item__expanded-area__right-area__info-item__value">$0.20</p>
        </div>
        <div class="payout-history-item__expanded-area__right-area__info-item">
          <p class="payout-history-item__expanded-area__right-area__info-item__label">Distributed</p>
          <p class="payout-history-item__expanded-area__right-area__info-item__value">$0.30</p>
        </div>
        <div class="payout-history-item__expanded-area__right-area__divider"></div>
        <div class="payout-history-item__expanded-area__right-area__footer">
          <div class="payout-history-item__expanded-area__right-area__footer__transaction"><a href="https://zkscan.io/explorer/transactions/0xreceipt1" target="_blank" rel="noreferrer noopener" class="payout-history-item__expanded-area__right-area__footer__transaction__link">
              Transaction
            </a>
            <shareicon-stub class="payout-history-item__expanded-area__right-area__footer__transaction__icon"></shareicon-stub>
          </div>
          <p class="payout-history-item__expanded-area__right-area__footer__total">$0.30</p>
        </div>
      </div>
    </div>
  </transition-stub>
</div>
`;
